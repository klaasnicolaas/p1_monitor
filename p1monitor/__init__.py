"""Asynchronous Python client for the P1 Monitor API."""

from .models import Phases, Settings, SmartMeter, WaterMeter
from .p1monitor import P1Monitor, P1MonitorConnectionError, P1MonitorError

__all__ = [
    "P1Monitor",
    "P1MonitorError",
    "P1MonitorConnectionError",
    "SmartMeter",
    "Settings",
    "Phases",
    "WaterMeter",
]
