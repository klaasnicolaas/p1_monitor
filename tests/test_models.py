"""Test the models."""

import pytest
from aiohttp import ClientSession
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from p1monitor import (
    P1Monitor,
    P1MonitorConnectionError,
    P1MonitorNoDataError,
    Phases,
    Settings,
    SmartMeter,
    WaterMeter,
)

from . import load_fixtures


async def test_smartmeter(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    p1monitor_client: P1Monitor,
) -> None:
    """Test request from a P1 Monitor device - SmartMeter object."""
    aresponses.add(
        "192.168.1.2",
        "/api/v1/smartmeter",
        "GET",
        aresponses.Response(
            text=load_fixtures("smartmeter.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    smartmeter: SmartMeter = await p1monitor_client.smartmeter()
    assert smartmeter == snapshot


async def test_phases(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    p1monitor_client: P1Monitor,
) -> None:
    """Test request from a P1 Monitor device - Phases object."""
    aresponses.add(
        "192.168.1.2",
        "/api/v1/status",
        "GET",
        aresponses.Response(
            text=load_fixtures("phases.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    phases: Phases = await p1monitor_client.phases()
    assert phases == snapshot


async def test_watermeter(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    p1monitor_client: P1Monitor,
) -> None:
    """Test request from a P1 Monitor device - WaterMeter object."""
    aresponses.add(
        "192.168.1.2",
        "/api/v2/watermeter/day",
        "GET",
        aresponses.Response(
            text=load_fixtures("watermeter.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    watermeter: WaterMeter = await p1monitor_client.watermeter()
    assert watermeter == snapshot


async def test_no_watermeter_data_new(aresponses: ResponsesMockServer) -> None:
    """Test no WaterMeter data from P1 Monitor device."""
    aresponses.add(
        "192.168.1.2",
        "/api/v2/watermeter/day",
        "GET",
        aresponses.Response(
            text=load_fixtures("no_data.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )

    async with ClientSession() as session:
        client = P1Monitor(host="192.168.1.2", session=session)
        with pytest.raises(P1MonitorNoDataError):
            await client.watermeter()


async def test_no_watermeter_data_old(aresponses: ResponsesMockServer) -> None:
    """Test no WaterMeter data from P1 Monitor device."""
    aresponses.add(
        "192.168.1.2",
        "/api/v2/watermeter/day",
        "GET",
        aresponses.Response(
            status=404,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )

    async with ClientSession() as session:
        client = P1Monitor(host="192.168.1.2", session=session)
        with pytest.raises(P1MonitorConnectionError):
            await client.watermeter()


async def test_settings(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    p1monitor_client: P1Monitor,
) -> None:
    """Test request from a P1 Monitor device - Settings object."""
    aresponses.add(
        "192.168.1.2",
        "/api/v1/configuration",
        "GET",
        aresponses.Response(
            text=load_fixtures("settings.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    settings: Settings = await p1monitor_client.settings()
    assert settings == snapshot
