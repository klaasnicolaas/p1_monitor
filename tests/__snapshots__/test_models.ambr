# serializer version: 1
# name: test_phases
  Phases(voltage_phase_l1=229.0, voltage_phase_l2=0.0, voltage_phase_l3=0.0, current_phase_l1=4.0, current_phase_l2=0.0, current_phase_l3=0.0, power_consumed_phase_l1=863, power_consumed_phase_l2=0, power_consumed_phase_l3=0, power_produced_phase_l1=0, power_produced_phase_l2=0, power_produced_phase_l3=0)
# ---
# name: test_settings
  Settings(gas_consumption_price=0.86687, energy_consumption_price_high=0.24388, energy_consumption_price_low=0.22311, energy_production_price_high=0.24388, energy_production_price_low=0.22311)
# ---
# name: test_smartmeter
  SmartMeter(gas_consumption=2289.967, energy_tariff_period=<EnergyTariff.LOW: 'low'>, power_consumption=935, energy_consumption_high=2996.141, energy_consumption_low=5436.256, power_production=0, energy_production_high=4408.947, energy_production_low=1575.502)
# ---
# name: test_watermeter
  WaterMeter(consumption_day=128.0, consumption_total=1640.399, pulse_count=128.0)
# ---
